safedelete/utils.py

```diff
@@ -1,23 +1,49 @@
-import itertools
+from itertools import chain, filterfalse
 
 from django.contrib.admin.utils import NestedObjects
 from django.db import router
 
+from .config import HAS_CASCADED_FIELD_NAME
 
-def related_objects(obj):
-    """ Return a generator to the objects that would be deleted if we delete "obj" (excluding obj) """
+
+def related_objects(obj, undelete_control=False):
+    """ Return a generator to the objects that would be deleted if we delete "obj" (excluding obj)
+
+    Args:
+        undelete_control: Include filter in flatten method to bypass elements controling undelete cascading.
+    """
 
     collector = NestedObjects(using=router.db_for_write(obj))
     collector.collect([obj])
 
+    def cascade_undelete_bypass(elem):
+        """ Test if elem should be bypassed by cascade undelete """
+
+        return (
+            elem != obj
+            and hasattr(elem, HAS_CASCADED_FIELD_NAME)
+            and not getattr(elem, HAS_CASCADED_FIELD_NAME)
+        )
+
+    def replace_if_cascade_child(elem):
+        """ Wraps each item in elem with tuples including their cascading child classes """
+
+        if collector.edges.get(elem):
+            return (elem, collector.edges[elem])
+        else:
+            return (elem,)
+
     def flatten(elem):
         if isinstance(elem, list):
-            return itertools.chain.from_iterable(map(flatten, elem))
+            if undelete_control:
+                elem = filterfalse(cascade_undelete_bypass, elem)
+            expanded_elem = chain.from_iterable(map(replace_if_cascade_child, elem))
```

Comment:
Wouldn't it be easier to filter after the flatten (and keep the `collector.nested()`)?

Reply:
collector.nested() over written test (`test_undelete_with_cascade_control_class_included`) would return this structure:
```
[
    <Author: Author object (3)>,
    [
        <Article: Article object (3)>,
        [
            <Press: Press object (1)>,
            <Section: Section object (1)>,
            <Section: Section object (2)>,
            [
                <Table: Table object (1)>
            ],
            <Section: Section object (3)>,
            [
                <Table: Table object (2)>,
                <Table: Table object (3)>
            ]
        ]
    ]
]
```
in the test case, Section object (2) shouldn't be restored neither every object inside the list after it (Table object (1) in this case).
After the flatten as it was written before, it's not possible to filter because when it returns Table object (1) we don't have information about its parent (if it was returned or not). Maybe a cleaner way would be overwriting the `collector.collect([obj])` to apply filter there.

Would you mind giving me some extra details if you had a better solution? Really appreciate your suggestions.

Thanks again.

Reply:
I was gonna say that we should get this information if `DELETED_BY_CASCADE_FIELD_NAME`, but you're right that we should not undelete children of an object which was not deleted by cascade.

It still seemed too complex for me so I tried simplifying it:

```python
def related_objects(obj, only_deleted_by_cascade=False):
    """ Return a generator to the objects that would be deleted if we delete "obj" (excluding obj)

    Args:
        only_deleted_by_cascade: Include filter in flatten method to bypass elements controling undelete cascading.
    """

    collector = NestedObjects(using=router.db_for_write(obj))
    collector.collect([obj])

    def flatten(elem):
        if isinstance(elem, list):
            return chain.from_iterable(map(flatten, elem))
        elif obj != elem and (not only_deleted_by_cascade or getattr(elem, DELETED_BY_CASCADE_FIELD_NAME)):
            return (elem,)
        return ()

    return flatten(collector.nested())
```

That made me realize that you expected `Table` to be undeleted in the `test_undelete_with_cascade_control_class_included` test. But `DELETED_BY_CASCADE_FIELD_NAME` has been unset in `Table`, so it should never be undeleted by cascade imo. What did you want to test exactly with `vars()[DELETED_BY_CASCADE_FIELD_NAME] = None`? I'm not sure if this is a behavior we should actually encourage.

Reply:
Thanks for your help over here. I was trying to test the behavior that you described in this comment https://github.com/makinacorpus/django-safedelete/pull/197#issuecomment-1032374114, when user don't want the additional field,... If you prefer I'll be removing the test and also this description part in SafeDeleteModel docstring:
```
        If this behavior isn't desired, class that inherits from SafeDeleteModel can override this
        attribute by setting it as None: overriding model class won't have its ``deleted_by_cascade``
        field and will be restored by cascading undelete even if it wasn't deleted by a cascade operation.
```

Reply:
`Table` objects in the test `test_undelete_with_cascade_control_class_included` (as you mentioned) were deleted by a cascading even with `DELETED_BY_CASCADE_FIELD_NAME` set as None (as expected). This was the reason that I used this approach to undelete it then. The code that you proposed changes this behavior... so, should I continue as you suggested? I'm just certifying the rules to keep us working on the same page.

Reply:
We expect objects having the `DELETED_BY_CASCADE_FIELD_NAME` attribute set to True to be undeleted in cascade. The logical things to me seems to be to ignore those which don't have it (equivalent to False).

I forgot you actually had documented that as well :+1: . We can keep it like that, just specify that models which don't have the field won't be undeleted instead.

Thanks for your work so far on this PR, we're almost there :+1: .

Reply:
Hey @Gagaro .I 'd like to let you know that I haven't found a cleaner way to write this filter yet. The solution you have proposed won't work if `<Section: Section object (2)> ` has childs with DELETED_BY_CASCADE_FIELD_NAME set to True. I 'm still trying to reach a better solution and I will cover this case also in the tests, maybe adding a new testing model with ForeignKey to section and DELETED_BY_CASCADE_FIELD_NAME=True.

Reply:
Do you mean because `<Section: Section object (2)>` has already been deleted before? If so, it should not be undeleted (and neither should its children). Its children should be undeleted only if `<Section: Section object (2)>` is undeleted directly (because it is the one which cause them to be deleted in the first place).

Reply:
You got it! That's exactly my point!

Reply:
```
def related_objects(obj, only_deleted_by_cascade=False):
    """ Return a generator to the objects that would be deleted if we delete "obj" (excluding obj)

    Args:
        only_deleted_by_cascade: Include filter in flatten method to bypass elements controling undelete cascading.
    """

    collector = NestedObjects(using=router.db_for_write(obj))
    collector.collect([obj])    

    def flatten(elem):
        if isinstance(elem, tuple):
            return elem
        elif obj == elem or not only_deleted_by_cascade or getattr(elem, DELETED_BY_CASCADE_FIELD_NAME):
            elem = [(elem,) if elem != obj else (), *collector.edges.get(elem, [])]
            return chain.from_iterable(map(flatten, elem))
        return ()

    return flatten(obj)
```
@Gagaro What do you think about the above code?

Reply:
Looks good to me!